src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1374,42 +1420,126 @@ RouterJob(Query *originalQuery, RelationRestrictionContext *restrictionContext,
 		return job;
 	}
 
-	pg_get_query_def(originalQuery, queryString);
-
 	if (originalQuery->commandType == CMD_SELECT)
 	{
-		task = CreateTask(ROUTER_TASK);
+		job->taskList = SingleShardSelectTaskList(originalQuery, relationShardList,
+												  placementList, shardId);
+	}
+	else if (isMultiShardModifyQuery)
+	{
+		job->taskList = MultiShardModifyTaskList(originalQuery, relationShardList,
+												 requiresMasterEvaluation);
 	}
 	else
 	{
-		DistTableCacheEntry *modificationTableCacheEntry = NULL;
-		char modificationPartitionMethod = 0;
+		job->taskList = SingleShardModifyTaskList(originalQuery, relationShardList,
+												  placementList, shardId);
+	}
 
-		modificationTableCacheEntry = DistributedTableCacheEntry(
-			updateOrDeleteRTE->relid);
-		modificationPartitionMethod = modificationTableCacheEntry->partitionMethod;
+	*multiShardModifyQuery = isMultiShardModifyQuery;
+	job->requiresMasterEvaluation = requiresMasterEvaluation;
+	return job;
+}
 
-		if (modificationPartitionMethod == DISTRIBUTE_BY_NONE &&
-			SelectsFromDistributedTable(rangeTableList))
+
+/*
+ * SingleShardSelectTaskList generates a task for single shard select query
+ * and returns it as a list.
+ */
+static List *
+SingleShardSelectTaskList(Query *query, List *relationShardList, List *placementList,
+						  uint64 shardId)
+{
+	Task *task = CreateTask(ROUTER_TASK);
+	StringInfo queryString = makeStringInfo();
+
+	pg_get_query_def(query, queryString);
+
+	task->queryString = queryString->data;
+	task->anchorShardId = shardId;
+	task->taskPlacementList = placementList;
+	task->relationShardList = relationShardList;
+
+	return list_make1(task);
+}
+
+
+/*
+ * MultiShardModifyTaskList generates task list for multi shard update/delete
+ * queries.
+ */
+static List *
+MultiShardModifyTaskList(Query *originalQuery, List *relationShardList,
+						 bool requiresMasterEvaluation)
+{
+	List *taskList = NIL;
+	ListCell *relationShardCell = NULL;
+	int taskId = 1;
+
+	foreach(relationShardCell, relationShardList)
+	{
+		RelationShard *relationShard = (RelationShard *) lfirst(relationShardCell);
+		List *relationShardList = list_make1(relationShard);
+		Query *copiedQuery = copyObject(originalQuery);
+		Task *task = CreateTask(MODIFY_TASK);
+		StringInfo shardQueryString = makeStringInfo();
+
+		if (!requiresMasterEvaluation)
 		{
-			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-							errmsg("cannot perform select on a distributed table "
-								   "and modify a reference table")));
+			UpdateRelationToShardNames((Node *) copiedQuery, relationShardList);
```

Comment:
We could move `Query *copiedQuery = copyObject(originalQuery);` here such that we don't have to copy the query tree unless we need it.